qiskit/extensions/quantum_initializer/state_preparation.py

```diff
@@ -0,0 +1,462 @@
+
+        Returns:
+            DAGCircuit: the circuit implementing the multiplexor's action
+        """
+        list_len = len(list_of_angles)
+        local_num_qubits = int(math.log2(list_len)) + 1
+
+        q = QuantumRegister(local_num_qubits)
+        circuit = QuantumCircuit(q, name="multiplex" + local_num_qubits.__str__())
+
+        lsb = q[0]
+        msb = q[local_num_qubits - 1]
+
+        # case of no multiplexing: base case for recursion
+        if local_num_qubits == 1:
+            circuit.append(target_gate(list_of_angles[0]), [q[0]])
+            return circuit
+
+        # calc angle weights, assuming recursion (that is the lower-level
+        # requested angles have been correctly implemented by recursion
+        angle_weight = np.kron([[0.5, 0.5], [0.5, -0.5]], np.identity(2 ** (local_num_qubits - 2)))
+
+        # calc the combo angles
+        list_of_angles = angle_weight.dot(np.array(list_of_angles)).tolist()
+
+        # recursive step on half the angles fulfilling the above assumption
+        multiplex_1 = self._multiplex(target_gate, list_of_angles[0 : (list_len // 2)], False)
+        circuit.append(multiplex_1.to_instruction(), q[0:-1])
+
+        # attach CNOT as follows, thereby flipping the LSB qubit
+        circuit.append(CXGate(), [msb, lsb])
+
+        # implement extra efficiency from the paper of cancelling adjacent
+        # CNOTs (by leaving out last CNOT and reversing (NOT inverting) the
+        # second lower-level multiplex)
+        multiplex_2 = self._multiplex(target_gate, list_of_angles[(list_len // 2) :], False)
+        if list_len > 1:
+            circuit.append(multiplex_2.to_instruction().reverse_ops(), q[0:-1])
+        else:
+            circuit.append(multiplex_2.to_instruction(), q[0:-1])
+
+        # attach a final CNOT
+        if last_cnot:
+            circuit.append(CXGate(), [msb, lsb])
+
+        return circuit
+
+
+def prepare_state(self, state, qubits=None):
+    r"""Prpare qubits in a specific state.
+
+    Args:
+        params (str or list or int):
+            * str: labels of basis states of the Pauli eigenstates Z, X, Y. See
+                :meth:`~qiskit.quantum_info.states.statevector.Statevector.from_label`.
+                Notice the order of the labels is reversed with respect to the qubit index to
+                be applied to. Example label '01' initializes the qubit zero to `|1>` and the
+                qubit one to `|0>`.
+            * list: vector of complex amplitudes to initialize to.
+            * int: an integer that is used as a bitmap indicating which qubits to initialize
+               to `|1>`. Example: setting params to 5 would initialize qubit 0 and qubit 2
+               to `|1>` and qubit 1 to `|0>`.
+        qubits (QuantumRegister or int):
+            * QuantumRegister: A list of qubits to be initialized [Default: None].
+            * int: Index of qubit to be initialized [Default: None].
```

Comment:
What happens if it's `None`?  The type hint is also inaccurate, since `None` is a valid input.

Reply:
I'm not sure how best to document this. If the number of qubits in the circ matches exactly the number needed to append the initialize instruction then the user doesn't have to pass anything for this arg. But if the circ is longer they do need to pass this arg to specify which qubits to append the instruction to.

Reply:
I think some form of the explanation you wrote is fine.  Like
> If given, a sequence of qubits or qubit specifiers to perform the initialisation on.  If not given or `None`, then the initialisation will be applied to all the qubits in the circuit.